File: samples/node-test/net-client.js
const net = require('net');
const pipeName = '\\\\.\\pipe\\sample';
const fs = require('fs-extra');
const cfg = fs.readJsonSync('net-cfg.json');

let client;
if (cfg.isPipe) {
  client = net.connect(pipeName, () => {
    // 'connect' listener
    console.log('connected to pipe server!');
    client.write('pipe world!\r\n');
  });
} else {
  client = net.connect(3000, 'localhost', () => {
    console.log('connected to tcp server!');
    client.write('tcp world!\r\n');
  });
}

client.on('data', (data) => {
  console.log('recv:', data.toString());

  // close connect
  client.end();
});

client.on('end', () => {
  console.log('disconnected from server');
});

client.on('error', (err) => {
  console.log('client error:', err);
});

File: samples/net-client/net-client.cpp
// net-client.cpp : Defines the entry point for the console application.
//

#include "stdafx.h"

string pipeName = R"(\\.\pipe\sample)";

class App : public Singleton<App> {
public:
  void run();

private:
  SocketPtr_t client_;
};

App& app = App::instance();

void App::run() {
  string jsonFile = path.join(path.dirname(process.execPath), "../../samples/node-test/net-cfg.json");
  json cfg = fs.readJsonSync(jsonFile);

  if (cfg["isPipe"].get<bool>()) {
    client_ = net.connect(pipeName, [this]() {
      cout << "pipe client connected to " << pipeName << endl;
      client_->write(Buffer("hello server, i am pipe client."));
    });
  }
  else {
    client_ = net.connect(3000, "127.0.0.1", [this]() {
      cout << "tcp client connected to " << pipeName << endl;
      client_->write(Buffer("hello server, i am tcp client."));
    });
  }

  client_->on<const Error&>("error", [](const Error& err) {
    cout << "client error: " << err.str() << endl;
  });

  client_->on<const Buffer&>("data", [](const Buffer& buf) {
    cout << "recv from svr: " << buf.toString() << endl;
  });

  nodecpp::run();
}

int _tmain(int argc, _TCHAR* argv[])
{
  app.run();
	return 0;
}

